### PR TITLE
fixup! trace2: collect Windows-specific process information

### PR DESCRIPTION
**Pull request by jeffhostetler**
Guard against infinite loop while computing the parent process hierarchy.

CreateToolhelp32Snapshot() is used to get a list of all processes on the
system.  Each process entry contains the process PID and PPID (alive at the
time of the snapshot).  We compute the set of ancestors of the current process
by repeated searches on this list.

Testing revealed that the snapshot can contain PPID cycles.  This causes an
infinite loop during the set construction and causes the git.exe command to
hang.

Testing found an instance where 3 processes were in a PPID cycle.  The
snapshot implied that each of these processes was its own great-grandparent.
This should not be possible unless a PID was recycled and just happened to
match up.

For full disclosure, the Windows "System Idle Process" has PID and PPID 0.
If it were to launch a Git command, it could cause a similar infinite loop.
Or more properly, if any ancestor of the current Git command has PPID 0, it
will appear to be a descendant of the idle process and trigger the problem.

This commit fixes both cases by maintaining a list of the PIDs seen during
the ancestor walk and stopping if a cycle is detected.

Additionally, code was added to truncate the search after a reasonable fixed
depth limit.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

### CODE REVIEW

**Comment by jeffhostetler** [comment]
This change are cherry-picked from my upstream submission: https://github.com/gitgitgadget/git/pull/108

File: compat/win32/trace2_win32_process_info.c [hunk 1]
```diff
@@ -38,16 +44,28 @@ static void get_processes(struct json_writer *jw, HANDLE hSnapshot)
...
+		/* Check for cycle in snapshot. (Yes, it happened.) */
+		for (k = 0; k < nr_pids; k++)
+			if (pid == pid_list[k]) {
+				jw_array_string(jw, "(cycle)");
+				return;
+			}
 
-	while (find_pid(pid, hSnapshot, &pe32)) {
-		jw_array_string(jw, pe32.szExeFile);
+		if (nr_pids == NR_PIDS_LIMIT) {
```
Comment (jrbriggs) [inline compat/win32/trace2_win32_process_info.c]: Paranoia nit: `if (nr_pids >= NR_PIDS_LIMIT)` ?

File: compat/win32/trace2_win32_process_info.c [hunk 2]
```diff
@@ -38,16 +44,28 @@ static void get_processes(struct json_writer *jw, HANDLE hSnapshot)
...
+	while (find_pid(pid, hSnapshot, &pe32)) {
+		/* Only report parents. Omit self from the JSON output. */
+		if (nr_pids)
+			jw_array_string(jw, pe32.szExeFile);
 
-	/* We only want parent processes, so skip self. */
-	if (!find_pid(pid, hSnapshot, &pe32))
-		return;
-	pid = pe32.th32ParentProcessID;
+		/* Check for cycle in snapshot. (Yes, it happened.) */
```
Comment (jrbriggs) [inline compat/win32/trace2_win32_process_info.c]: Do we need to check for cycles or should we just give up once we exceed NR_PIDS_LIMIT?

File: compat/win32/trace2_win32_process_info.c (hunk 1 above)
Comment (jeffhostetler) [reply compat/win32/trace2_win32_process_info.c]: true, but we'd already have a buffer overrun at that point.

File: compat/win32/trace2_win32_process_info.c (hunk 2 above)
Comment (jeffhostetler) [reply compat/win32/trace2_win32_process_info.c]: yea, i could go either way here.  my way lets us see "(cycle)" vs "(truncate)" in the Kusto data.
i thought that might be interesting.

Comment (jamill) [reply compat/win32/trace2_win32_process_info.c]: If we check for a cycle (which we know happens), we will avoid having `NR_PIDS_LIMIT` lines of repeated entries every time it does happen...

**Pull request merged**